pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -344,6 +344,7 @@ func ApplyRevision(set *apps.StatefulSet, revision *apps.ControllerRevision) (*a
 	if err != nil {
 		return nil, err
 	}
+	clone.Reset()
 	err = json.Unmarshal(patched, clone)
```

Comment:
- The documentation(https://golang.org/pkg/encoding/json/#Unmarshal) doesn't mention, that that interface (`clone` here) needs to be initialized. The ideal unmarshal should use the patched and populate all fields in clone and , leave others to default value.
- also the clone := set.DeepCopy() at line 342 doesnt seem necessary since the next line is a read only operation(i know this is old code)

-One way to verify if this indeed fixes the problem, would be to use two versions of ApplyRevision, one without the reset and one with the reset and see that the first one always fails and the second one always passes, ideally i would checkin both tests, but at the very least we can validate if this is the right fix.

Reply:
* The interface needs to be initialized. Run simple program below:
```go
package main

import (
        "fmt"
        "encoding/json"
)

func main() {
        type demo struct {
                foo *int `json:"foo"`
        }
        bar := 1
        d := demo{foo: &bar}
        // if `json.Unmarshal` uses default value, d.foo == nil
        if err := json.Unmarshal([]byte("{}"), &d); err != nil {
                panic(err)
        }
        fmt.Println(*d.foo)
}
```
the program does not panic, and the output is 1

* Removed the `clone := set.DeepCopy()`